### PR TITLE
[LangRef] Specify NaN behavior more precisely

### PR DESCRIPTION
**Pull request by RalfJung**
The current docs  don't say anything about the possible set of values returned by a NaN-producing operation. However, there are some coding patterns where such guarantees are strongly desired, such as NaN boxing (as used e.g. in the Spidermonkey JS engine). And in fact the set of possible payloads that can be observed today when compiling code via LLVM is fairly limited (based on what optimizations and backends and hardware currently do) -- but without a documented guarantee, languages compiling to LLVM cannot really rely on this.

There was a long discussion about the exact shape that the guarantees should take [on Discourse](https://discourse.llvm.org/t/stronger-floating-point-nan-guarantees/72165). This PR transcribes what I perceived as the consensus, plus some further clarifications: "unless specified otherwise", returned NaNs have a non-deterministic sign, and the payload is generally either the preferred (all-zero) payload or propagates the payload of one of the inputs.

### CODE REVIEW

File: llvm/docs/LangRef.rst
Comment (programmerjake) [inline llvm/docs/LangRef.rst:3403]: floating point bits *do* mutate arbitrarily on 32-bit x86:
here, all I do is bitcast `0x7f800001` to `float`, return it from a function, then bitcast back to an integer -- that changes it to `0x7fc00001`.
https://clang.godbolt.org/z/sjaq55q3G

Comment (RalfJung) [reply llvm/docs/LangRef.rst:3403]: Yes, that's a bug in the x86-32 backend / calling convention.

Given that this bug is hard-coded in the calling convention, I wonder if it should be called out here. But the main point if this sentence is that LLVM itself will not do anything that would mutate FP values. In fact this bug here leads to miscompilations because LLVM itself *assumes* that FP values don't change -- see https://github.com/llvm/llvm-project/issues/44218.

File: llvm/docs/LangRef.rst
Comment (Muon) [inline llvm/docs/LangRef.rst:3412]: Nit: the signaling bit is not part of the payload.

Comment (RalfJung) [reply llvm/docs/LangRef.rst:3412]: According to which definition?

In [wasm](https://webassembly.github.io/spec/core/syntax/values.html#floating-point), they definitely consider it to be part of the payload
> NaN values have a payload that describes the mantissa bits in the underlying [binary representation](https://webassembly.github.io/spec/core/exec/numerics.html#aux-fbits).

File: llvm/docs/LangRef.rst
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:3397]: The `denormal-fp-math` attribute does affect whether subnormals are preserved, although I suspect it is somewhat inconsistently applied at the moment.

File: llvm/docs/LangRef.rst
Comment (jcranmer-intel) [reply llvm/docs/LangRef.rst:3412]: > According to which definition?

Per IEEE 754-2019, section 6.2.1:
> In the preferred encoding just described, a signaling NaN shall be quieted by setting d1 to 1, leaving the remaining bits of T unchanged. Bits d2 d3 … dp−1 of the trailing significand field contain the encoding of the payload

**Review by jcranmer-intel — COMMENTED** [review]
It is worth calling out somewhere that qNaN/sNaN are defined using the convention that leading bit = 0 is sNaN, 1 is qNaN, and systems that do not adhere to this convention (e.g., old MIPS) are unsupported.

Also probably worth calling out that things like `bitcast` do not count as floating-point operations.

File: llvm/docs/LangRef.rst
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3412]: Ah, looks like wasm picked a different definition, that's unfortunate (and the fact that the IEEE spec is not publicly accessible). I'll update to the IEEE definition then. I should probably just briefly explicitly state what exactly we consider the payload, and that we assume that the leading bit is a "quiet" bit (1 = qNaN).

File: llvm/docs/LangRef.rst
Comment (jyknight) [reply llvm/docs/LangRef.rst:3405]: Those 3 shouldn't be considered floating point math operations (as the term is used here and before). We definitely need to make it clear in their definitions -- it has implications beyond just the NaN behavior.

In particular, they all ought to be specified as: "This operation never raises a floating-point exception, and the result is an exact copy of the input, other than the sign bit. It is not considered a floating-point math operation, but rather, bit-manipulation which operates on floating-point values."

File: llvm/docs/LangRef.rst
Comment (jyknight) [reply llvm/docs/LangRef.rst:3403]: I do think we should at least mention the major ways in which LLVM backends violate the spec:
- x86-32 with SSE2 enabled may implicitly convert floating-point values returned from a function to x86_fp80 and back; such conversions may signal a floating-point exception, or modify the value.
- x86-32 without SSE2 may do such conversions just about _anywhere_.
- older MIPS versions use the opposite polarity for the sNaN vs qNaN bit, and LLVM does not correctly represent this (this is at least potentially fixable, unlike the above).

File: llvm/docs/LangRef.rst
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3405]: The paragraph above says that FP exceptions cannot be observed anyway, so it seems confusing/misleading to now talk about some operations not raising FP exceptions.

File: llvm/docs/LangRef.rst
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3397]: I made the text mention this attribute now, is that better?

File: llvm/docs/LangRef.rst
Comment (RalfJung) [reply llvm/docs/LangRef.rst]: I'm mentioning `strictfp` now, I hope what I am saying makes sense. :)

File: llvm/docs/LangRef.rst
Comment (RalfJung) [reply llvm/docs/LangRef.rst:3425]: I am explicitly calling out bitcasts and fnet, fabs, copysign as not being affected by the NaN non-determinism now.

File: llvm/docs/LangRef.rst
Comment (RalfJung) [reply llvm/docs/LangRef.rst]: I added a paragraph with backend problems and linked to relevant issues. Is there an issue for the "x86-32 return value problem"?

File: llvm/docs/LangRef.rst
Comment (RalfJung) [reply llvm/docs/LangRef.rst]: I updated the text to treat the quiet bit as being separate from the payload.

File: llvm/docs/LangRef.rst
Comment (RalfJung) [inline llvm/docs/LangRef.rst:3408]: This wasn't really about the "FP environment" any more so I moved this all into a new section on NaN values.

**Comment by RalfJung** [comment]
Regarding fmin/fmax on x86, @Muon pointed out that LLVM currently simply calls the underlying libm functions to implement these operations. The C standard has nothing to say about the case where both inputs are a NaN. My local manpage says

> If both arguments are NaN, a NaN is returned.

This is not very specific and could permit arbitrary NaN payloads. I don't know if that actually happens in practice though.

So I'm still unsure about whether we need to say that fmin/fmax on x86 can produce arbitrary NaN payloads.

File: llvm/docs/LangRef.rst
Comment (programmerjake) [inline llvm/docs/LangRef.rst:3400]:
```suggestion
to undefined behavior. The ``strictfp`` and ``denormal-fp-math`` attributes as
```

Comment (RalfJung) [reply llvm/docs/LangRef.rst]: *oops*

File: llvm/docs/LangRef.rst
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:3413]: Nit: QNaN, not QNan here.

File: llvm/docs/LangRef.rst
Comment (jcranmer-intel) [reply llvm/docs/LangRef.rst:3397]: I think this works for now.

File: llvm/docs/LangRef.rst
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:3462]: Nit: mention "for some calling conventions" here?

File: llvm/docs/LangRef.rst
Comment (jcranmer-intel) [reply llvm/docs/LangRef.rst:3425]: I would still add "unless otherwise specified", because you have cases like `llvm.canonicalize` which is an FP operation that can arbitrarily mutate NaN payloads but requires the output to be qNaN, even if the input is sNaN.

**Review by jcranmer-intel — APPROVED** [review]

File: llvm/docs/LangRef.rst
Comment (nunoplopes) [inline llvm/docs/LangRef.rst:3435]: Given that optimizations must be correct for all targets, this rule is equal to saying that any qnan may be returned. Can we simplify the rule to say just that?

Comment (RalfJung) [reply llvm/docs/LangRef.rst:3435]: It's not equivalent, no. When compiling code to x86 I'd like to get a guarantee that LLVM will not introduce any non-preferred NaNs. It's a target-specific guarantee. It's not useful for optimizations, but it is useful for people compiling LLVM IR for a specific target.

Comment (RalfJung) [reply llvm/docs/LangRef.rst:3435]: Formally speaking, the Abstract Machine becomes parameterized by this set of target-specific NaNs. For Alive I guess that could be tricky since to be fully accurate it has to verify that the transformation is correct for every possible target-NaN set.

Comment (nunoplopes) [reply llvm/docs/LangRef.rst:3435]: Ok, optimizations need to be correct for different kinds of targets, some for which the set is empty up to a full set. Let's see which optimizations break :)

File: llvm/docs/LangRef.rst
Comment (nunoplopes) [inline llvm/docs/LangRef.rst:3431]: just to confirm, even in these rules that copy the input we still have a non-deterministic sign?
So the return value is (nondet-sign) ++ 0xFF ++ fraction as determined by these rules

Comment (RalfJung) [reply llvm/docs/LangRef.rst:3431]: Yes, that is the intention. The choice of sign and (quiet + payload) is independent. (But quiet and payload are correlated.)

I'm open to suggestions for how to word this more clearly.

File: llvm/docs/LangRef.rst
Comment (jyknight) [reply llvm/docs/LangRef.rst]: I'm not aware of one in LLVM, only in rust, so I filed a new one, #66803.

**Comment by nunoplopes** [comment]
I've a question regarding fpext & fptrunc: what's the payload we want there? We can't just copy the inputs.

**Comment by nunoplopes** [comment]
We need a fix for these instructions, so that SNaN can passthrough.

**Comment by jcranmer-intel** [comment]
> I've a question regarding fpext & fptrunc: what's the payload we want there? We can't just copy the inputs.

`fpext` and `fptrunc` correspond to the IEEE 754 operation **convertFormat**, so the rules related to SNaN handling need to apply (i.e., it may or may not be quieting input SNaN). IEEE 754 furthermore recommends NaN propagation as follows:

> Conversion of a quiet NaN from a narrower format to a wider format in the same radix, and then back to
the same narrower format, should not change the quiet NaN payload in any way except to make it
canonical.

The natural way to implement this is to designate certain bits to chop, but I don't know what the full suite of implementations here is. In x86, the afflicted bits are the LSB, so `fpext` on `0xffffffff` results in `0xffffffffe0000000`, and I think `APFloat` is following the same rules here, but I'm not certain all NaN-propagating implementations would follow the same rules.

**Comment by nunoplopes** [comment]
> > I've a question regarding fpext & fptrunc: what's the payload we want there? We can't just copy the inputs.
> 
> `fpext` and `fptrunc` correspond to the IEEE 754 operation **convertFormat**, so the rules related to SNaN handling need to apply (i.e., it may or may not be quieting input SNaN). IEEE 754 furthermore recommends NaN propagation as follows:
> 
> > Conversion of a quiet NaN from a narrower format to a wider format in the same radix, and then back to
> > the same narrower format, should not change the quiet NaN payload in any way except to make it
> > canonical.
> 
> The natural way to implement this is to designate certain bits to chop, but I don't know what the full suite of implementations here is. In x86, the afflicted bits are the LSB, so `fpext` on `0xffffffff` results in `0xffffffffe0000000`, and I think `APFloat` is following the same rules here, but I'm not certain all NaN-propagating implementations would follow the same rules.

I've implemented what you suggested and the 4 cases above now verify.
I have zero regressions on the LLVM test suite with the proposed semantics! 🚀

**Comment by Muon** [comment]
I'm pretty sure the significand lives in the least significant bits on every single architecture. For format conversions I'm pretty sure everyone just truncates/zero extends.

**Comment by programmerjake** [comment]
> I'm pretty sure the significand lives in the least significant bits on every single architecture. For format conversions I'm pretty sure everyone just truncates/zero extends.

IEEE754 specifies fp->fp format conversions should use the programmer-chosen rounding mode, for LLVM for f32 <-> f64 that's always round to nearest, ties to even

**Comment by Muon** [comment]
Ah, I meant for NaN handling.

File: llvm/docs/LangRef.rst
Comment (RalfJung) [reply llvm/docs/LangRef.rst]: > I'd like to see some text added to the specification sections for the fneg/fabs/copysign operations. 

Done.

**Comment by RalfJung** [comment]
So for fpext/fptruncate, what does this mean? Should we add something like this?

`fpext`:
> If ``value`` is a NaN, then the result will be a NaN with the same sign, with the quiet/signaling bit either preserved or reset to ``1``, and with the payload extended by extension with
least-significant zeroes.

`fptruncate`:
> If ``value`` is a NaN, then the result will be a NaN with the same sign, with the quiet/signaling bit either preserved or reset to ``1``, and with the payload truncated by removing the
least significant bits.

Or is that not correct and the payload may be reset to the preferred payload as well?
Are we sufficiently sure that this is correct for all hardware? It seems to be stricter than what IEEE mandates. The alternative would be to say that the payload is arbitrary.

**Comment by programmerjake** [comment]
IIRC RISC-V converts NaN payloads to the canonical qNaN when converting between fp formats.

**Comment by RalfJung** [comment]
So, like this?

`fpext`:
> If ``value`` is a NaN, then the result will be a NaN with the same sign, with the quiet/signaling bit non-deterministically either preserved or reset to ``1``, and with the payload non-deterministically being all-zero or the payload of ``value`` extended with trailing (least-significant) zeroes.

`fptruncate`:
> If ``value`` is a NaN, then the result will be a NaN with the same sign, with the quiet/signaling bit non-deterministically either preserved or reset to ``1``, and with the payload non-deterministically being all-zero or the payload of ``value`` truncated by removing the least significant bits.

Though maybe it's easier to just say

`fpext`+`fptruncate`:
> If ``value`` is a NaN, then the result will be a NaN with the same sign, with the quiet/signaling bit non-deterministically either preserved or reset to ``1``, and with the payload non-deterministically chosen.

Or even

`fpext`+`fptruncate`:
> If ``value`` is a NaN, then the result will be a NaN with the quiet/signaling bit non-deterministically either preserved or reset to ``1``, and with the sign and payload non-deterministically chosen.

**Comment by jyknight** [comment]
The section "What happens to the payload when the precision is converted?" in [This 2018 paper](https://grouper.ieee.org/groups/msc/ANSI_IEEE-Std-754-2019/background/nan-propagation.pdf) claimed that all known implementations preserve the upper NaN bits when truncating, and append zeros when extending.

Of course, that's only the case for the ISAs which propagate NaN payloads at all (so, not RISC-V and not WebAssembly). On RISCV, the sign is always positive and the NaN canonical. On Wasm, the sign is non-deterministic, and the payload follows Wasm's usual "nondeterministic result if given non-canonical input NaNs, canonical NaN with nondeterministic sign otherwise" rule.

Given that these are the exact same issues that arise for all other FP instructions, I think these instructions should use our normal rules, just clarified to say how the bits are copied *if* they happen to be copied.

Thus, I suggest the spec should say something like this for `fptrunc`,
> NaN values follow the usual :ref:\`NaN behaviors \<floatnan\>\`, except that \_if\_ a NaN payload is copied from the input, then the low order bits of the NaN payload which cannot fit in the resulting type are discarded.

And for `fpext`:
> NaN values follow the usual :ref:\`NaN behaviors \<floatnan\>\`, except that \_if\_ a NaN payload is copied from the input, then it is copied to the high order bits of the resulting payload, and the remaining low order bits are zero.

**Comment by RalfJung** [comment]
I have adjusted this a bit to clarify what is mean by "_if_ a NaN payload is copied from the input", and pushed that to the PR.

**Comment by RalfJung** [comment]
@nikic @jcranmer-intel @efriedma-quic what are the next steps towards landing this PR?

**Comment by nunoplopes** [comment]
> @nikic @jcranmer-intel @efriedma-quic what are the next steps towards landing this PR?

It's approved; just click the button "squash and merge" (or ask us to do it for you) 🚀

**Comment by RalfJung** [comment]
A bunch of changes happened since it got approved though. I'm not familiar with how LLVM handles that situation.

Also I don't have that button. :)

**Comment by efriedma-quic** [comment]
> A bunch of changes happened since it got approved though. I'm not familiar with how LLVM handles that situation.

If you think the changes are significant enough to require another review, please point out the specific parts.  Otherwise, we can consider the patch ready.

> Also I don't have that button. :)

From the developer policy: "If you do not have commit access, please let people know during the review and someone should commit it on your behalf."

**Review by jyknight — APPROVED** [review]

**Comment by jyknight** [comment]
Before we merge, though: can you rewrite the commit description (by editing the title&first comment) to accurately describe the current situation?

Thx!

**Comment by jyknight** [comment]
> I've updated the PR description. The PR title still seems fine?

Thanks. I tweaked the title slightly.

> Should I squash the commits?

No, you shouldn't -- squash happens _during_ the merge (via the button labeled "Squash and Merge"). I will press said button now.

**Comment by RalfJung** [comment]
Great, thanks a lot. :)

**Pull request merged**
